### ISSUE TITLE
bundling sources

### ISSUE THREAD

**Issue by topepo** [astamm/nloptr#34]
I've repeatedly had issues with the sources at [http://ab-initio.mit.edu/nlopt](http://ab-initio.mit.edu/nlopt) being unavailable. It seems to consistently break travis builds. 

Is it possible to bundle the sources in the R package to avoid this? I'm sure that this has been thought of or previously discussed but `lme4` and a few other widely downloaded packages depend on `nloptr`. 

As an alternative, I might be able to get RStudio to host the sources so that we would have a more reliable link. Perhaps that could be a fallback download source in your build system.

**Comment by jspiewak** [comment]
I am actually [looking](https://github.com/rstudio/shinyapps-package-dependencies/issues/97) into this right now.

The download URL should be updated to fetch from GitHub. I'll open a pull request here.

**Comment by topepo** [comment]
Thanks. Let me know if you need a hand with anything.

**Comment by jyypma** [comment]
If NLopt is installed and this can be detected by pkg-config in the nloptr
configure script, then the source code of NLopt will not be downloaded.
Instead, nloptr will link to the existing NLopt installation.

The request of adding the NLopt source to the nloptr library has come up
before before. I have been hesitant to add the source code because (1) I
don't know how this would work with licensing, and  (2) I don't want to
take credit for the NLopt code, since I did not write it.

I'll reconsider adding the NLopt source to the nloptr package the next time
I will try to make a submission to CRAN, which I plan to do end of August.


On 17 Jul 2017 6:50 p.m., "Max Kuhn" <notifications@github.com> wrote:

Thanks. Let me know if you need a hand with anything.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/jyypma/nloptr/issues/34#issuecomment-315809174>, or mute
the thread

**Comment by topepo** [comment]
>  I have been hesitant to add the source code because (1) I
don't know how this would work with licensing, and  (2) I don't want to
take credit for the NLopt code, since I did not write it.

I think that @jspiewak 's PR will solve the issue that I'm having. It's more about the lack of reliability of the current url.

**Comment by eddelbuettel** [comment]
> I've repeatedly had issues with the sources at http://ab-initio.mit.edu/nlopt being unavailable. It seems to consistently break travis builds.

There is a *very* easy fix: install `libnlopt-dev` under Travis, and nloptr will use it.   Also speeds up the compilation tremendously as you no longer need to builds libnlopt's code.

**Comment by topepo** [comment]
> There is a very easy fix: install libnlopt-dev under Travis, and nloptr will use it. Also speeds up the compilation tremendously as you no longer need to builds libnlopt's code.

I'll do that. However, I've had a lot of issues when I'm teaching a class to brand new R users.  Either from binaries for `nloptr` not being available on the CRAN repos or the users not being able to build on OS X.

**Comment by eddelbuettel** [comment]
That's a slightly shifted goal post, but could you not pre-preare binaries and host a local repo if you desire binaries?   

[ I don't disagree that macOS has issues with compiling package but that is a different can of worms.  I just wanted to point out that _for Travis_ the issue is indeed very easily fixable. ]

**Comment by topepo** [comment]
Not a problem... I didn't really say anything originally about the binaries.

It looks like a moot point. From jyypma/nloptr#35

> In the previous commits, I have included the NLopt source code as subdirectory of src. If no pre-installed NLopt version can be found on the system during the configuration of nloptr, the included NLopt code is configured and compiled. NLopt is no longer downloaded during configuration which solves the issue raised in this pull request.

@jyypma Should I close this?

**Comment by topepo** [comment]
It failed again on travis 
```
trying URL 'http://ab-initio.mit.edu/nlopt/nlopt-2.4.2.tar.gz'
Error in download.file(url = "http://ab-initio.mit.edu/nlopt/nlopt-2.4.2.tar.gz", :
 cannot open URL 'http://ab-initio.mit.edu/nlopt/nlopt-2.4.2.tar.gz'
Execution halted
```
I'll try Dirk's suggestion

**Comment by eddelbuettel** [comment]
There are two main advantages to it (and I so need to blog on this):
- a binary saves you time from compiling
- a binary can make installation easier

Both hold here for nloptr after we mod'ed the `configure` script to use it.  

Of course, this mostly only helps those on platform rich enough to actually have lots of binaries (in decent shape).  An under-appreciated feature of Travis is its Ubuntu basis:  there are over 25k binary packages in the distro alone, a bunch of very useful PPAs (ie when you need or want newest g++ or clang) --- and in the R case the 5k r-cran-* binary packages provided by Michael Rutter.

**Comment by jyypma** [comment]
Closed by f8cba33a8c052e2d7a10bfa5b84650d670fe4374
